### PR TITLE
Andrew/stencil

### PR DESCRIPTION
**Pull request by exspeed**
Press 1 to activate stencil.

TODO (later): have naming standard for the shaders 
Update ReadMe

### CODE REVIEW

File: src/InputManager.cpp [hunk 1]
```diff
@@ -32,11 +33,24 @@ void InputManager::MouseInput()
...
 	{
 		glfwSetWindowShouldClose(window, GL_TRUE);
 	}
+	for(int i = 0; i < scene->selected.size() && i < 9; i++)
+	{
+		static int oldState = GLFW_RELEASE;
+		int newState = glfwGetKey(window, GLFW_KEY_0+i+1);
+		if(newState == GLFW_RELEASE && oldState == GLFW_PRESS)
+		{
+			scene->selected[i] = scene->selected[i] ? 0 : 1;
```
Comment (ammunoz) [inline src/InputManager.cpp]: Easier to just make selected a property of Geometries. That way we don't have to maintain two vectors.

File: src/Scene.cpp [hunk 2]
```diff
@@ -46,24 +30,78 @@ void Scene::ClearGeometries()
...
+			glDrawArrays(geometries[i]->renderMode, 0, geometries[i]->vertices.size());
+		}
+		else
+		{
+			RenderStencil(geometries[i]);
+		}
 	}
-
+	
+	glEnable(GL_DEPTH_TEST);
```
Comment (ammunoz) [inline src/Scene.cpp]: Does this not need to turn on before we draw?

File: src/main.cpp [hunk 3]
```diff
@@ -29,32 +29,40 @@ int main(int argc, char *argv[])
...
+
+	
+	InputManager inputManager(window, camera, scene);
 
 	CheckGLErrors();
 
+	glEnable(GL_DEPTH_TEST);
+	glEnable(GL_STENCIL_TEST);
+	glStencilFunc(GL_NOTEQUAL, 1, 0xFF);
+	glStencilOp(GL_KEEP, GL_KEEP, GL_REPLACE);
```
Comment (ammunoz) [inline src/main.cpp]: These are better off in GLHelpers::Initialize instead of main.

Also, are the enables here even necessary? Don't these get called by Render and/or RenderStencil in each frame anyways?

File: src/Scene.cpp [hunk 4]
```diff
@@ -46,24 +30,78 @@ void Scene::ClearGeometries()
...
+	program->setMat4("MVP", MVP);
+	
+		// bind our shader program and the vertex array object containing our
+		// scene geometry, then tell OpenGL to draw our geometry
+	glBindVertexArray(geometry->vertexArray);
+	glDrawArrays(geometry->renderMode, 0, geometry->vertices.size());
+	
+	glStencilFunc(GL_NOTEQUAL, 1, 0xFF);
+	glStencilMask(0x00);
+	glDisable(GL_DEPTH_TEST);
```
Comment (ammunoz) [inline src/Scene.cpp:90]: It is confusing to know when we need GL_DEPTH_TEST as it is getting turned on and off in different places.

Is it possible to localize the enabling and disabling of it? Better to have redundant but understandable code. If not, please add brief comments so readers can follow along.

A possible solution is to enable GL_DEPTH_TEST during Scene::Render() and disabling it at the end of the same function as it seems necessary to be enabled in Scene::Render() but disabled in Scene::RenderStencil().

File: shaders/fragment.glsl [hunk 5]
```diff
@@ -16,4 +16,5 @@ void main(void)
 {
 	// write colour output without modification
 	FragmentColour = vec4(Colour, 0);
+
```
Comment (ammunoz) [inline shaders/fragment.glsl]: Unnecessary change

File: src/InputManager.cpp [hunk 6]
```diff
@@ -32,11 +33,24 @@ void InputManager::MouseInput()
 	cursorY = ypos;
 	
 }
+
+#include <iostream>
```
Comment (ammunoz) [inline src/InputManager.cpp]: This should get moved up top

**Review by ammunoz — CHANGES_REQUESTED** [review]

File: src/main.cpp (hunk 3 above)
Comment (exspeed) [reply src/main.cpp]: d

File: src/InputManager.cpp (hunk 1 above)
Comment (exspeed) [reply src/InputManager.cpp]: update geometry to have get/set. Since scene had geometry as private also made getter/setters for geometries.

File: src/main.cpp (hunk 3 above)
Comment (exspeed) [reply src/main.cpp]: moved to GLHelpers::Initialize

removed 
+	glEnable(GL_DEPTH_TEST);
+	glEnable(GL_STENCIL_TEST);

other 2 is necessary

File: src/Scene.cpp (hunk 4 above)
Comment (exspeed) [reply src/Scene.cpp:90]: remove some unecessary DepthTest should be clear now.

File: src/Geometry.cpp [hunk 7]
```diff
@@ -115,6 +117,17 @@ void Geometry::Load() const
 	glBindBuffer(GL_ARRAY_BUFFER, 0);
 }
 
+void Geometry::SetSelectedGeometry()
+{
+	selected = selected ? 0 : 1;
```
Comment (ammunoz) [inline src/Geometry.cpp]: selected = !selected;

File: src/Geometry.cpp [hunk 8]
```diff
@@ -115,6 +117,17 @@ void Geometry::Load() const
 	glBindBuffer(GL_ARRAY_BUFFER, 0);
 }
 
+void Geometry::SetSelectedGeometry()
```
Comment (ammunoz) [inline src/Geometry.cpp]: Better to call it ToggleSelect.

SetSelectedGeometry() implies a setter.

File: src/Scene.cpp [hunk 9]
```diff
@@ -42,28 +25,91 @@ void Scene::ClearGeometries()
...
+	programOutline->use();
+	
+	model = glm::scale (glm::vec3(1.1f, 1.1f, 1.1f));
+	MVP =camera->ProjectionMatrix* camera->ViewMatrix * model;
+	programOutline->setMat4("MVP", MVP);
+	
+	glBindVertexArray(geometry->vertexArray);
+	glDrawArrays(geometry->renderMode, 0, geometry->vertices.size());
+	
+	glEnable(GL_DEPTH_TEST);
```
Comment (ammunoz) [inline src/Scene.cpp:101]: Much better

File: src/main.cpp [hunk 10]
```diff
@@ -44,32 +44,34 @@ int main(int argc, char *argv[])
...
 		glfwSwapBuffers(window);
 
 		glfwPollEvents();
 	}
 
 	// Clean up allocated resources before exit
-	scene.ClearGeometries();
+	scene->ClearGeometries();
 	delete(camera);
 	delete(inputManager);
```
Comment (ammunoz) [inline src/main.cpp]: Have to delete scene

File: src/main.cpp [hunk 11]
```diff
@@ -44,32 +44,34 @@ int main(int argc, char *argv[])
...
-	// Create Scene
-	Scene scene("shaders/vertex.glsl", "shaders/fragment.glsl", camera);
-	scene.AddGeometry(geometry);
+	Shader program("shaders/vertex.glsl","shaders/fragment.glsl");
+	Shader programOutline("shaders/vertex.glsl","shaders/outline.frag");
+	Scene* scene = new Scene(&program, &programOutline, camera);
+	scene->AddGeometry(geometry);
 
-	CheckGLErrors();
+	inputManager = new InputManager(window, camera, scene);
```
Comment (ammunoz) [inline src/main.cpp:57]: Remove new keyword from here.

inputManager = &InputManager(window, camera, scene);

InputManager should never go out of scope. If it does, program should crash.

If there is a good reason otherwise, just make sure to delete(inputManager) at the end.

File: src/Scene.cpp [hunk 12]
```diff
@@ -42,28 +25,91 @@ void Scene::ClearGeometries()
 	geometries.clear();
 }
 
+int Scene::GetGeometriesSize()
+{
+	return geometries.size();
+}
+
+void Scene::SetSelectedGeometry(int i)
```
Comment (ammunoz) [inline src/Scene.cpp]: Maybe call this Scene::ToggleSelectedGeometry(..) as well

**Review by ammunoz — APPROVED** [review]
Fix these and merge

File: src/main.cpp (hunk 11 above)
Comment (exspeed) [reply src/main.cpp:57]: inputManager is a global pointer. Keeping it as pointer with new, will make inputManager static though (private in main).
